content/blog/2019-10-08-remote-read-meets-streaming.md

```diff
@@ -0,0 +1,313 @@
+// be sent for previous one.
+message ChunkedReadResponse {
+  repeated prometheus.ChunkedSeries chunked_series = 1;
+}
+
+// ChunkedSeries represents single, encoded time series.
+message ChunkedSeries {
+  // Labels should be sorted.
+  repeated Label labels = 1 [(gogoproto.nullable) = false];
+  // Chunks will be in start time order and may overlap.
+  repeated Chunk chunks = 2 [(gogoproto.nullable) = false];
+}
+```
+
+As you can see the frame does not include raw samples anymore. That's the second improvement we did: We send in the message
+samples batched in chunks (see [this video](https://www.youtube.com/watch?v=b_pEevMAC3I) to learn more about chunks), 
+which are exactly the same chunks we store in the TSDB. 
+
+We ended up with the following server algorithm:
+
+1. Parse request.
+1. Select metrics from TSDB.
+1. For all series:
+    * For all samples:
+        * Encode into chunks
+            * if the frame is >= 1MB; break
+  * Marshal `ChunkedReadResponse` message.
+  * Snappy compress
+  * Send the message   
+
+You can find full design [here](https://docs.google.com/document/d/1JqrU3NjM9HoGLSTPYOvR217f5HBKBiJTqikEB9UiJL0/edit#).
+
+## Benchmarks
+
+How does this performance of this new approach compare to the old solution?
+
+Let's compare remote read characteristics between Prometheus `2.12.0` and `2.13.0`. As the initial results presented 
+at the beginning of this article, I was using Prometheus as a server, and Thanos sidecar as a client of the remote read.
+I was invoking testing remote read request by running gRPC call against Thanos sidecar using `grpcurl`.
+Test was performed from my laptop (Lenovo X1 16GB, i7 8th) with Kubernetes in docker (using [kind](https://github.com/kubernetes-sigs/kind)).
+
+The data was artificially generated, and represents highly dynamic 10k series (worst case scenario).
+
+The full test bench is available in [thanosbench repo](https://github.com/thanos-io/thanosbench/blob/master/benchmarks/remote-read/README.md).
+
+### Memory
+
+![Prometheus 2.12.0: Heap-only allocations of single read 8h of 10k series](/assets/blog/2019-10-08/10series8hours-2.12-allocs.png)
+
+![Prometheus 2.13.0: Heap-only allocations of single read 8h of 10k series](/assets/blog/2019-10-08/10series8hours-2.13-allocs.png)
+
+Reducing memory was the key item we aimed for with our solution. Prometheus instead of allocating GBs of memory, buffers 
+roughly 50MB during the whole request, whereas for Thanos there is only a marginal memory use. Thanks to the streamed
+Thanos gRPC StoreAPI, sidecar is now a very simple proxy.
+
+Additionally, I tried different time ranges and number of series, but as expected I kept seeing
+maximum of 50MB in allocations for Prometheus and nothing really visible for Thanos. This proves that our remote read
+uses **constant memory per request** allowing straightforward capacity planning against user traffic, with help of the concurrency limit. 
```

Comment:
I don't think it's actually constant, due to the indexing.

Reply:
True, how to put it then?

Reply:
constant memory no matter how many samples there are or similar